### PR TITLE
Introduce folding argument in ZNE mitigation

### PR DESCRIPTION
**Pull request by cosenal**
**Context:**
We'd like to extend Zero-noise extrapolation (ZNE) with different folding techniques. For more details, see #755 and  https://mitiq.readthedocs.io/en/stable/guide/zne-3-options.html#unitary-folding.

**Description of the Change:**
In this PR a `folding` argument is added both to the ZNE mitigation routine in Catalyst API/frontend/MLIR dialect.
At the moment the argument is passed throughout the stack from the frontend all the way to the MLIR dialect. However it is mute, in the sense that it's not being used to change the behavior of the MLIR ZNE transform. 

**Benefits:**
This is an incremental change to discuss design and to accomodate follow-up pull requests where the argument will be used to branch off the ZNE mitigation code for different unitary folding techniques.

**Possible Drawbacks:**
The argument is left as `global` by default and it's not yet used in the MLIR transform, so the current behavior is unchanged. If one tries to use different value for the new argument, exceptions are raised. 

**Related GitHub Issues:**
This is an incremental change towards #755.

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @cosenal! I've pinged @rmoyard to take a look

**Comment by cosenal** [comment]
Thanks @josh146. Thanks to @dime10, I resolved the errors, and this approach now works 🎉

I haven't publicized the PR yet, because I wanted to clean it up and write tests, but before I do that, it may be a good idea if you or Romain take a preliminary look and give me feedback on the approach.

**Comment by dime10** [comment]
> Thanks, @josh146. Thanks to @dime10, I have now resolved the errors, and this approach now works 🎉
> 
> I haven't publicized the PR yet, because I wanted to clean it up and write tests, but before I do that, it may be a good idea if you or Romain take a preliminary look and give me feedback on the approach.

Thanks @cosenal, this looks great to me!

**Comment by cosenal** [comment]
@dime10 @rmoyard @WrathfulSpatula This PR is now ready for review (I don't have permission to formally request reviewers on this repo).

File: mlir/include/Mitigation/IR/MitigationOps.td
Comment (dime10) [inline mlir/include/Mitigation/IR/MitigationOps.td:23]: Minor, but it would be nice to keep these sorted (at least roughly in terms of grouping like `IR` and `Interfaces`, unfortunately we don't have a tool that does this for tablegen files).

File: mlir/lib/Mitigation/IR/MitigationOps.cpp
Comment (dime10) [inline mlir/lib/Mitigation/IR/MitigationOps.cpp:30]: Why were these moved down? Not a big deal but a bit unusual since the files should normally be generated without assuming particular namespace commands.

File: mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp
Comment (dime10) [inline mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp:49]: The C++ compiler will probably raise a warning about the unused variable, maybe this avoids it:
```suggestion
    op.getFolding();
```

**Review by dime10 — COMMENTED** [review]
Thanks for the PR, I only have minor suggestions :)

File: frontend/catalyst/api_extensions/error_mitigation.py
Comment (rmoyard) [inline frontend/catalyst/api_extensions/error_mitigation.py:59]: Can you document the different options only when we can actually use them?

File: frontend/catalyst/api_extensions/error_mitigation.py
Comment (rmoyard) [inline frontend/catalyst/api_extensions/error_mitigation.py:153]: You can add a test for this case :+1:

File: mlir/lib/Mitigation/IR/MitigationDialect.cpp
Comment (rmoyard) [inline mlir/lib/Mitigation/IR/MitigationDialect.cpp:71]: Is it necessary here? As you already added it in intialize?

File: mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp
Comment (rmoyard) [reply mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp:49]: Is it necessary to be added in this PR?

**Review by rmoyard — COMMENTED** [review]
Thanks great PR :+1: I have added some suggestions

File: mlir/lib/Mitigation/IR/MitigationOps.cpp
Comment (cosenal) [reply mlir/lib/Mitigation/IR/MitigationOps.cpp:30]: I moved them down for coherence with https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Quantum/IR/QuantumOps.cpp#L30-L32

File: mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp
Comment (cosenal) [reply mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp:49]: No, it was a leftover from debugging. I will remove as this belongs to a follow-up PR.

**Review by cosenal — COMMENTED** [review]
@rmoyard @dime10 all comments addressed, PTAL

**Review by rmoyard — APPROVED** [review]
It looks good to me :100: Please just wait for @dime10 feedback before merging.

**Review by dime10 — APPROVED** [review]
Great 💯

**Pull request merged**
